plugins/openai_plugin.py

```diff
@@ -133,143 +178,160 @@ def request_http(self, query: dict, user_id: int, test_end_time: float = 0):
-
-                    # First non empty token is the first token
-                    if not result.first_token_time and token != "":
-                        result.first_token_time = time.time()
-
-                    # If the current token time is outside the test duration, record the total tokens received before
-                    # the current token.
-                    if (
-                        not result.output_tokens_before_timeout
-                        and time.time() > test_end_time
-                    ):
-                        result.output_tokens_before_timeout = len(tokens)
-
-                    tokens.append(token)
-
-                    # Last token comes with finish_reason set.
-                    if message.get("choices", [])[0].get("finish_reason", None):
-                        result.stop_reason =  message["choices"][0]["finish_reason"]
-
-                        # If test duration timeout didn't happen before the last token is received, 
-                        # total tokens before the timeout will be equal to the total tokens in the response.
-                        if not result.output_tokens_before_timeout:
-                            result.output_tokens_before_timeout = result.output_tokens
-
-                except KeyError:
-                    logger.exception("KeyError, unexpected response format in line: %s", line)
+                recv_time = time.time() # Record time asap
+                # Only record lines with data
+                if line:
+                    logger.debug("response line: %s", line)
+                    resps.append(dict(
+                        time = recv_time,
+                        data = line
+                    ))
+            # Full response received
+            result.end_time = time.time()
         except requests.exceptions.ChunkedEncodingError as err:
             result.end_time = time.time()
             result.error_text = repr(err)
-            result.output_text = "".join(tokens)
-            result.output_tokens = len(tokens)
+            #result.output_text = "".join([])
+            result.output_tokens = len(resps)
             if response is not None:
                 result.error_code = response.status_code
             logger.exception("ChunkedEncodingError while streaming response")
             return result
 
+        ###########################################
+        # DO NOT CALL time.time BEYOND THIS POINT #
+        ###########################################
+
+        # If no data was received return early
+        if not resps:
+            result.output_tokens = 0
+            result.error_code = response.status_code
+            return result
+
+        # Check for end of request marker
+        if resps[-1]['data'] == b"data: [DONE]":
+            result.end_time = resps[-1]['time']
+            resps.pop() # Drop the end indicator
+        else:
+            # TODO This signals that the request is incomplete
+            pass
```

Comment:
In the previous implementation, we don't do anything if the end marker is missing. Do we want to have some sort of error state for this?

Reply:
My concern with hardcoding the end request marker is that it is not a defined openai API (as far as I know). Could different openai server have their own implementation of end request markers? I remember not seeing this marker when hitting the openai endpoint exposed by the llama.cpp server. 

Not saying that we need to fix it in this PR and although hitting this issue is not very likely as most often we use vLLM but something we need to keep an eye on.

Reply:
IIRC we don't support llama.cpp due to other API differences.